src/header.rs

```diff
@@ -16,47 +16,210 @@ use other;
     pub prefix: [u8; 155],
-    _rest: [u8; 12],
+    pub pad: [u8; 12],
+}
+
+/// Representation of the header of an entry in an archive
+#[repr(C)]
+#[allow(missing_docs)]
+pub struct GnuHeader {
+    pub name: [u8; 100],
+    pub mode: [u8; 8],
+    pub uid: [u8; 8],
+    pub gid: [u8; 8],
+    pub size: [u8; 12],
+    pub mtime: [u8; 12],
+    pub cksum: [u8; 8],
+    pub typeflag: [u8; 1],
+    pub linkname: [u8; 100],
+
+    // GNU format
+    pub magic: [u8; 6],
+    pub version: [u8; 2],
+    pub uname: [u8; 32],
+    pub gname: [u8; 32],
+    pub dev_major: [u8; 8],
+    pub dev_minor: [u8; 8],
+    pub atime: [u8; 12],
+    pub ctime: [u8; 12],
+    pub offset: [u8; 12],
+    pub longnames: [u8; 4],
+    pub unused: [u8; 1],
+    pub sparse: [GnuSparseHeader; 4],
+    pub isextended: [u8; 1],
+    pub realsize: [u8; 12],
+    pub pad: [u8; 17],
+}
+
+#[repr(C)]
+#[allow(missing_docs)]
+pub struct GnuSparseHeader {
+    pub offset: [u8; 12],
+    pub numbytes: [u8; 12],
 }
 
 impl Header {
-    /// Creates a new blank ustar header ready to be filled in
+    /// Creates a new blank ustar header ready to be filled in.
+    ///
+    /// Note that the returned header is by default in the GNU format.
     pub fn new() -> Header {
-        let mut header: Header = unsafe { mem::zeroed() };
-        // Flag this header as a UStar archive
-        header.ustar = *b"ustar\0";
-        header.ustar_version = *b"00";
+        let mut header = Header { bytes: [0; 512] };
+        header.set_gnu();
         return header
     }
 
+    fn cast<T>(&self) -> &T {
+        assert_eq!(mem::size_of_val(self), mem::size_of::<T>());
+        unsafe { &*(self as *const Header as *const T) }
+    }
+
+    fn cast_mut<T>(&mut self) -> &mut T {
+        assert_eq!(mem::size_of_val(self), mem::size_of::<T>());
+        unsafe { &mut *(self as *mut Header as *mut T) }
+    }
+
     fn is_ustar(&self) -> bool {
-        &self.ustar[..5] == b"ustar"
+        let ustar = self.cast::<UstarHeader>();
+        ustar.magic[..] == b"ustar\0"[..] && ustar.version[..] == b"00"[..]
+    }
+
+    fn is_gnu(&self) -> bool {
+        let ustar = self.cast::<UstarHeader>();
+        ustar.magic[..] == b"ustar "[..] && ustar.version[..] == b" \0"[..]
+    }
+
+    /// View this archive header as a raw "old" archive header.
+    ///
+    /// This view will always succeed as all archive header formats will fill
+    /// out at least the fields specified in the old header format.
+    pub fn as_old(&self) -> &OldHeader {
+        self.cast()
+    }
+
+    /// Same as `as_old`, but the mutable version.
+    pub fn as_old_mut(&mut self) -> &mut OldHeader {
+        self.cast_mut()
+    }
+
+    /// Flags this header as an "old header"
+    ///
+    /// Note that this is a destructive operation and may corrupt fields that
+    /// have been previously set. It is recommended to call this operation
+    /// **first** when creating new headers.
+    pub fn set_old(&mut self) {
```

Comment:
If the operation needs to be done at first, why not make a new constructor like `new_old`? And remove this function?